tcmur_cmd_handler.c

```diff
@@ -628,26 +584,58 @@ static int handle_passthrough(struct tcmu_device *dev,
 	return ret;
 }
 
-/*
- * try to passthrough the command if handler supports command passthrough.
- * note that TCMU_NOT_HANDLED is returned when a tcmur handler does not
- * handle a passthrough command, but since we call ->handle_cmd via
- * async_call_command(), ->handle_cmd can finish in the callers context
- * (asynchronous handler) or work queue context (synchronous handlers),
- * thus we'd need to check if ->handle_cmd handled the passthough command
- * here as well as in handle_passthrough_cbk().
- */
+bool tcmur_handler_is_passthrough_only(struct tcmur_handler *rhandler)
+{
+	if (rhandler->write || rhandler->read || rhandler->flush)
+		return false;
+
+	return true;
+}
+
 int tcmur_cmd_handler(struct tcmu_device *dev, struct tcmulib_cmd *cmd)
 {
-	int ret;
+	int ret = TCMU_NOT_HANDLED;
 	struct tcmulib_handler *handler = tcmu_get_dev_handler(dev);
 	struct tcmur_handler *rhandler = handler->hm_private;
+	uint8_t *cdb = cmd->cdb;
+	struct iovec *iovec = cmd->iovec;
+	size_t iov_cnt = cmd->iov_cnt;
+	uint32_t block_size = tcmu_get_dev_block_size(dev);
+	off_t offset = block_size * tcmu_get_lba(cdb);
 
-	if (rhandler->handle_cmd) {
-		ret = handle_passthrough(dev, rhandler, cmd);
-		if ((ret == TCMU_ASYNC_HANDLED) || (ret != TCMU_NOT_HANDLED))
-			return ret;
```

Comment:
I am not sure what you mean. handle_passthrough is called from tcmur_cmd_handler.c. It is called for all handlers that use the tcmu-runner daemon. file_optical uses passthrough for all its commands for example. Is there a bug where it is not getting called?

I am also not sure what you mean by making it ceph specific. translate/report are called from the upper generic layer SCSI command processing aren't they? The upper level generic SCSI code in tcmu-runner processing the SCSI parts in alua.c.  Any handler can hook into the callbacks to implement any low level specifics like how they can for read/write/flush. For example, gluster is working on something similar to fence out other nodes during failover/failback. You could implement a drbd handler and for its transition callout you would do something like the drbdadm resrouce commands.  Like rbd, for these devices, we have to call into the handler for failover/failback to do low level operations like make sure commands from other nodes are not still running when switching nodes.

If you do not implement a transition/report callout and still wanted ALUA or your handler can do active/active and/or do not need any low level helper, you can setup ports group with targetcli, and tcmu-runner will report that ALUA info from the generic SCSI handler and it call just works for the handler.